lib/net.c

```diff
@@ -1776,9 +1776,192 @@ start_network_stage2(void *arg)
 }
 
 static int
-drop_privileges(void)
+copy_amb_to_inh(const char *block, cap_t cap_p)
+{
+	cap_value_t i;
+	for (i = 0; i <= CAP_LAST_CAP; i++) {
+		int ret;
+
+		int value = cap_get_ambient(i);
+		if (value < 0) {
+			G_LOG(WARNING, "%s (%s): capability %u not supported by the running kernel: %s\n",
+				__func__, block, i, strerror(errno));
+			continue;
+		}
+
+		ret = cap_set_flag(cap_p, CAP_INHERITABLE, 1, &i,
+			value ? CAP_SET : CAP_CLEAR);
+		if (ret != 0) {
+			G_LOG(ERR, "%s (%s): could not set CAP_INHERITABLE to %u for capability %u: %s\n",
+				__func__, block, value ? CAP_SET : CAP_CLEAR,
+				i, strerror(errno));
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static void
+log_proc_caps(const char *block, const char *context)
+{
+	cap_t cap_p = cap_get_proc();
+	char *cap_output, *amb_output;
+	int ret;
+
+	if (cap_p == NULL) {
+		G_LOG(ERR, "%s: cannot get capabilities for process when asked by block %s: %s\n",
+			__func__, block, strerror(errno));
+		return;
+	}
+
+	cap_output = cap_to_text(cap_p, NULL);
+	if (cap_output == NULL) {
+		G_LOG(ERR, "%s: cannot get text string of capabilities when asked by block %s: %s\n",
+			__func__, block, strerror(errno));
+		goto proc;
+	}
+
+	/* Log ambient capabilities. */
+	cap_clear(cap_p);
+	ret = copy_amb_to_inh(block, cap_p);
+	if (ret < 0)
+		goto cap;
+
+	amb_output = cap_to_text(cap_p, NULL);
+	if (amb_output == NULL) {
+		G_LOG(ERR, "%s: cannot get text string of ambient capabilities when asked by block %s: %s\n",
+			__func__, block, strerror(errno));
+		goto cap;
+	}
+
+	G_LOG(DEBUG, "%s: %s: %s\n\t(ambient): %s\n",
+		block != NULL ? block : "Main",
+		context, cap_output, amb_output);
+
+	cap_free(amb_output);
+cap:
+	cap_free(cap_output);
+proc:
+	cap_free(cap_p);
+}
+
+int
+needed_caps(const char *block, int ncap, const cap_value_t *caps)
+{
+	cap_t cap_p;
+	int ret;
+
+	/* No capablities are needed when run as root. */
+	if (config.pw_uid == 0)
+		return 0;
+
+	log_proc_caps(block, "Capabilities before setting");
+
+	cap_p = cap_init();
+	if (cap_p == NULL) {
+		G_LOG(ERR, "%s (%s): could not create a capability state in working storage: %s\n",
+			__func__, block, strerror(errno));
+		return -1;
+	}
+
+	if (ncap > 0) {
+		ret = cap_set_flag(cap_p, CAP_PERMITTED, ncap, caps, CAP_SET);
+		if (ret != 0) {
+			G_LOG(ERR, "%s (%s): could not set CAP_PERMITTED for %d capabilities: %s\n",
+				__func__, block, ncap, strerror(errno));
+			goto free;
+		}
+
+		ret = cap_set_flag(cap_p, CAP_EFFECTIVE, ncap, caps, CAP_SET);
+		if (ret != 0) {
+			G_LOG(ERR, "%s (%s): could not set CAP_EFFECTIVE for %d capabilities: %s\n",
+				__func__, block, ncap, strerror(errno));
+			goto free;
+		}
+	}
+
+	ret = cap_set_proc(cap_p);
+	if (ret != 0) {
+		G_LOG(ERR, "%s (%s): could not set capabilities for process: %s\n",
+			__func__, block, strerror(errno));
+		goto free;
+	}
+free:
+	cap_free(cap_p);
+
+	if (ret < 0)
+		return ret;
+
+	ret = cap_reset_ambient();
```

Comment:
Let's test for `CAP_AMBIENT_SUPPORTED()` before trying a reset. Otherwise, Gatekeeper cannot run on kernels that don't support ambient capabilities.